### PR TITLE
mds: switch to use projected inode instead

### PR DESCRIPTION
**Pull request by lxbsz**
When the client send a sync request to MDS even the client has the
Xx caps, for example, the MDS will revoke the Xx caps back and journal
new changes.

If the client try to create a sub-directory the MDS was always using
the non-projected inode of parent to inherit the gid/uid/mode metadatas,
if the journal log still haven't been flushed and applied yet, it will
always get the old metadatas.

Fixes: https://tracker.ceph.com/issues/56010
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by lxbsz** [comment]
> The commit message says MDS uses non-projected inode of parent to inherit gid/uid/mode metadata. Looking at the code, it's only the setgid bit that was incorrectly inherited from non-projected inode, right?

No, both the `gid` and `setgid bit` were incorrectly inherited.

**Comment by lxbsz** [comment]
> Although I can not reproduce this locally, the issue seems to be real and the fix looks good to me.

I can reproduce it locally every time with latest `testing` branch.

**Comment by ajarr** [comment]
> > The commit message says MDS uses non-projected inode of parent to inherit gid/uid/mode metadata. Looking at the code, it's only the setgid bit that was incorrectly inherited from non-projected inode, right?
> 
> No, both the `gid` and `setgid bit` were incorrectly inherited.

I'm sorry I missed to add `gid`. But `uid` is not inherited, right?

**Comment by lxbsz** [comment]
> 



> > > The commit message says MDS uses non-projected inode of parent to inherit gid/uid/mode metadata. Looking at the code, it's only the setgid bit that was incorrectly inherited from non-projected inode, right?
> > 
> > 
> > No, both the `gid` and `setgid bit` were incorrectly inherited.
> 
> I'm sorry I missed to add `gid`. But `uid` is not inherited, right?

From the posix:

```
The directory's user ID shall be set to the process' effective user ID.  The directory's group ID shall be set to the group ID of the parent directory or to the effective group ID of the process. Implementations shall provide a way to initialize the directory's  group  ID  to  the group  ID  of  the  parent directory. Implementations may, but need not, provide an implementation-defined way to initialize the directory's group ID to the effective group ID of the calling process.
```
The uid should always be the process' effective user ID for new creating directories.

File: src/mds/Server.cc [hunk 1]
```diff
@@ -3372,18 +3372,20 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
   _inode->truncate_seq = 1; /* starting with 1, 0 is kept for no-truncation logic */
 
   CInode *diri = dir->get_inode();
+  const CInode::mempool_inode *pip = diri->get_projected_inode().get();
```
Comment (vshankar) [inline src/mds/Server.cc:3375]: nit: use `auto`.

Comment (vshankar) [reply src/mds/Server.cc:3375]: also, any reason not to work with just a `CInode *`?

Comment (lxbsz) [reply src/mds/Server.cc:3375]: > nit: use `auto`.

Sure, will fix it.

> also, any reason not to work with just a `CInode *`?

Sorry, didn't get it here. What do you mean `CInode *` ? Do you mean `diri->get_inode()` ?

**Review by kotreshhr — APPROVED** [review]
lgtm

**Review by vshankar — APPROVED** [review]

**Review by batrick — APPROVED** [review]
Great analysis and fix. Well done!

**Comment by batrick** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2022-July-04

**Comment by vshankar** [comment]
jenkins test make check

**Pull request merged**